File: MidtermLabWork3/JAVA/dataset.java
package MidtermLabWork3.JAVA;

import java.io.*;
import java.util.*;

public class dataset {
    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        
        // Prompt user for dataset file path
        System.out.print("Enter the dataset file path: ");
        String filePath = scanner.nextLine();
        
        try {
            // Read CSV file
            List<String[]> records = readCSV(filePath);
            
            // MP01: Display total number of records
            System.out.println("Total number of records: " + records.size());
            
            // MP02: Display the first 10 rows
            System.out.println("\nFirst 10 rows:");
            for (int i = 0; i < Math.min(10, records.size()); i++) {
                System.out.println(Arrays.toString(records.get(i)));
            }
            
            // MP20: Convert to JSON
            String json = convertToJSON(records);
            System.out.println("\nJSON format:");
            System.out.println(json);
            
        } catch (FileNotFoundException e) {
            System.out.println("Error: File not found. Please check the file path.");
        } catch (IOException e) {
            System.out.println("Error: Unable to read the file.");
        } catch (Exception e) {
            System.out.println("Error: Invalid data or parsing issue.");
        }
        
        scanner.close();
    }
    
    // Function to read CSV file
    // This function reads the CSV file and returns a list of string arrays representing each row
    private static List<String[]> readCSV(String filePath) throws IOException {
        List<String[]> records = new ArrayList<>();
        String line;
        boolean headerFound = false;
        
        try (BufferedReader br = new BufferedReader(new FileReader(filePath))) {
            // Skip lines until header is found
            while ((line = br.readLine()) != null) {
                if (line.contains("Candidate")) {
                    headerFound = true;
                    break;
                }
            }
            if (!headerFound) {
                throw new IOException("Header not found in CSV file.");
            }
            
            // Read data rows
            while ((line = br.readLine()) != null) {
                if (!line.trim().isEmpty()) {
                    String[] fields = parseCSVLine(line);
                    records.add(fields);
                }
            }
        }
        return records;
    }
    
    // Simple CSV parser that handles quoted fields
    // This is a basic parser; for complex CSV, use a library like OpenCSV
    private static String[] parseCSVLine(String line) {
        List<String> fields = new ArrayList<>();
        boolean inQuotes = false;
        StringBuilder field = new StringBuilder();
        
        for (char c : line.toCharArray()) {
            if (c == '"') {
                inQuotes = !inQuotes;
            } else if (c == ',' && !inQuotes) {
                fields.add(field.toString().trim());
                field.setLength(0);
            } else {
                field.append(c);
            }
        }
        fields.add(field.toString().trim());
        
        return fields.toArray(new String[0]);
    }
    
    // Function to convert records to JSON
    // This creates a simple JSON array of objects
    private static String convertToJSON(List<String[]> records) {
        StringBuilder json = new StringBuilder();
        json.append("[\n");
        
        for (int i = 0; i < records.size(); i++) {
            String[] record = records.get(i);
            json.append("  {\n");
            json.append("    \"Candidate\": \"").append(record.length > 0 ? escapeJSON(record[0]) : "").append("\",\n");
            json.append("    \"Type\": \"").append(record.length > 1 ? escapeJSON(record[1]) : "").append("\",\n");
            json.append("    \"Column1\": \"").append(record.length > 2 ? escapeJSON(record[2]) : "").append("\",\n");
            json.append("    \"Exam\": \"").append(record.length > 3 ? escapeJSON(record[3]) : "").append("\",\n");
            json.append("    \"Language\": \"").append(record.length > 4 ? escapeJSON(record[4]) : "").append("\",\n");
            json.append("    \"Exam Date\": \"").append(record.length > 5 ? escapeJSON(record[5]) : "").append("\",\n");
            json.append("    \"Score\": \"").append(record.length > 6 ? escapeJSON(record[6]) : "").append("\",\n");
            json.append("    \"Result\": \"").append(record.length > 7 ? escapeJSON(record[7]) : "").append("\",\n");
            json.append("    \"Time Used\": \"").append(record.length > 8 ? escapeJSON(record[8]) : "").append("\"\n");
            json.append("  }");
            if (i < records.size() - 1) {
                json.append(",");
            }
            json.append("\n");
        }
        
        json.append("]");
        return json.toString();
    }
    
    // Helper to escape JSON strings
    private static String escapeJSON(String str) {
        return str.replace("\\", "\\\\").replace("\"", "\\\"");
    }
}

File: MidtermLabWork3/JAVASCRIPT/dataset.js
const fs = require('fs');
const readline = require('readline');

// Create readline interface for user input
const rl = readline.createInterface({
    input: process.stdin,
    output: process.stdout
});

// Function to prompt user for dataset file path
// This function asks the user to enter the path to the CSV file
rl.question('Enter the dataset file path: ', (filePath) => {
    try {
        // Read the CSV file synchronously
        // This reads the entire file into a string for processing
        const csvData = fs.readFileSync(filePath, 'utf8');

        // Parse CSV data into records
        // This function splits the CSV into lines and parses each line
        const records = parseCSV(csvData);

        // MP01: Display total number of records
        // This shows the count of data rows (excluding header)
        console.log(`Total number of records: ${records.length}`);

        // MP02: Display the first 10 rows
        // This loops through the first 10 records and prints them
        console.log('\nFirst 10 rows:');
        for (let i = 0; i < Math.min(10, records.length); i++) {
            console.log(records[i]);
        }

        // MP20: Convert to JSON format
        // This converts the array of records to a JSON string
        const jsonData = convertToJSON(records);
        console.log('\nJSON format:');
        console.log(jsonData);

    } catch (error) {
        // Handle possible errors during file reading or processing
        if (error.code === 'ENOENT') {
            console.log('Error: File not found. Please check the file path.');
        } else {
            console.log('Error: Unable to read or process the file.');
        }
    }

    // Close the readline interface
    rl.close();
});

// Function to parse CSV data
// This function takes the raw CSV string and converts it to an array of objects
function parseCSV(csvString) {
    const lines = csvString.split('\n');
    const records = [];
    let headerFound = false;

    // Skip lines until header is found
    for (let line of lines) {
        if (line.includes('Candidate')) {
            headerFound = true;
            break;
        }
    }

    if (!headerFound) {
        throw new Error('Header not found in CSV file.');
    }

    // Process data rows
    for (let i = lines.indexOf(lines.find(line => line.includes('Candidate'))) + 1; i < lines.length; i++) {
        const line = lines[i].trim();
        if (line) {
            const record = parseCSVLine(line);
            records.push(record);
        }
    }

    return records;
}

// Function to parse a single CSV line
// This handles quoted fields and commas within quotes
function parseCSVLine(line) {
    const fields = [];
    let field = '';
    let inQuotes = false;

    for (let i = 0; i < line.length; i++) {
        const char = line[i];
        if (char === '"') {
            inQuotes = !inQuotes;
        } else if (char === ',' && !inQuotes) {
            fields.push(field.trim());
            field = '';
        } else {
            field += char;
        }
    }
    fields.push(field.trim());

    return fields;
}

// Function to convert records to JSON
// This creates a JSON array of objects with predefined keys
function convertToJSON(records) {
    const jsonArray = records.map(record => ({
        "Candidate": record[0] || "",
        "Type": record[1] || "",
        "Column1": record[2] || "",
        "Exam": record[3] || "",
        "Language": record[4] || "",
        "Exam Date": record[5] || "",
        "Score": record[6] || "",
        "Result": record[7] || "",
        "Time Used": record[8] || ""
    }));

    return JSON.stringify(jsonArray, null, 2);
}
